File: src/util/exec.ts
import { spawn, spawnSync, type SpawnOptions } from "node:child_process";
import { log } from "./logger.js";

// Promise-based exec wrappers that surface stdout/stderr cleanly.

export interface ExecResult {
  code: number;
  stdout: string;
  stderr: string;
}

export async function run(
  cmd: string,
  args: string[],
  opts: SpawnOptions & { capture?: boolean; quiet?: boolean } = {},
): Promise<ExecResult> {
  const { capture = false, quiet = false, ...spawnOpts } = opts;
  return new Promise<ExecResult>((resolve) => {
    const child = spawn(cmd, args, {
      stdio: capture ? ["ignore", "pipe", "pipe"] : quiet ? ["ignore", "ignore", "ignore"] : "inherit",
      ...spawnOpts,
    });
    let stdout = "";
    let stderr = "";
    child.stdout?.on("data", (d) => (stdout += d.toString()));
    child.stderr?.on("data", (d) => (stderr += d.toString()));
    child.on("error", (err) => {
      log.debug(`spawn error for ${cmd}: ${err.message}`);
      resolve({ code: 127, stdout, stderr: stderr + err.message });
    });
    child.on("close", (code) => resolve({ code: code ?? 0, stdout, stderr }));
  });
}

export function which(bin: string): string | null {
  const PATH = process.env.PATH || "";
  const isWin = process.platform === "win32";
  const exts = isWin ? (process.env.PATHEXT || ".EXE;.CMD;.BAT").split(";") : [""];
  const sep = isWin ? ";" : ":";
  const fs = require("node:fs") as typeof import("node:fs");
  const path = require("node:path") as typeof import("node:path");
  for (const dir of PATH.split(sep)) {
    if (!dir) continue;
    for (const ext of exts) {
      const p = path.join(dir, bin + ext);
      try {
        if (fs.statSync(p).isFile()) return p;
      } catch {
      }
    }
  }
  return null;
}

export function whichAny(bins: string[]): { bin: string; path: string } | null {
  for (const b of bins) {
    const p = which(b);
    if (p) return { bin: b, path: p };
  }
  return null;
}

export function trySync(cmd: string, args: string[]): ExecResult {
  const r = spawnSync(cmd, args, { encoding: "utf8" });
  return {
    code: r.status ?? 127,
    stdout: r.stdout ?? "",
    stderr: r.stderr ?? r.error?.message ?? "",
  };
}

File: internal/util/exec.go
package util

import (
	"bytes"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// ExecResult mirrors the TS { code, stdout, stderr } shape.
type ExecResult struct {
	Code   int
	Stdout string
	Stderr string
}

// RunOptions controls stdio handling for Run.
type RunOptions struct {
	Capture bool
	Quiet   bool
	Cwd     string
	Env     []string
}

// Run executes a command; Capture pipes stdio, Quiet discards it, else inherit.
func Run(cmd string, args []string, opts RunOptions) ExecResult {
	c := exec.Command(cmd, args...)
	if opts.Cwd != "" {
		c.Dir = opts.Cwd
	}
	if opts.Env != nil {
		c.Env = append(os.Environ(), opts.Env...)
	}
	var outBuf, errBuf bytes.Buffer
	if opts.Capture {
		c.Stdout = &outBuf
		c.Stderr = &errBuf
	} else if opts.Quiet {
		// discard
	} else {
		c.Stdout = os.Stdout
		c.Stderr = os.Stderr
		c.Stdin = os.Stdin
	}
	err := c.Run()
	res := ExecResult{Stdout: outBuf.String(), Stderr: errBuf.String()}
	if err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			res.Code = ee.ExitCode()
		} else {
			// spawn failure (ENOENT etc.)
			res.Code = 127
			res.Stderr += err.Error()
		}
		return res
	}
	res.Code = 0
	return res
}

// Which finds an executable on PATH, honoring PATHEXT on Windows.
func Which(bin string) string {
	pathEnv := os.Getenv("PATH")
	var exts []string
	sep := ":"
	if IsWin {
		sep = ";"
		pe := os.Getenv("PATHEXT")
		if pe == "" {
			pe = ".EXE;.CMD;.BAT"
		}
		exts = strings.Split(pe, ";")
	} else {
		exts = []string{""}
	}
	for _, dir := range strings.Split(pathEnv, sep) {
		if dir == "" {
			continue
		}
		for _, ext := range exts {
			p := filepath.Join(dir, bin+ext)
			if fi, err := os.Stat(p); err == nil && !fi.IsDir() {
				return p
			}
		}
	}
	return ""
}

// WhichAny returns the first found bin and its path.
func WhichAny(bins []string) (string, string) {
	for _, b := range bins {
		if p := Which(b); p != "" {
			return b, p
		}
	}
	return "", ""
}
